aiida/orm/implementation/sqlalchemy/querybuilder/main.py

```diff
@@ -169,117 +202,75 @@ def get_session(self) -> Session:
+            result = query.count()
+        return result
 
     def first(self, data: QueryDictType) -> Optional[List[Any]]:
-        query = self.update_query(data)
-        try:
+        with self.use_query(data) as query:
             result = query.first()
-        except Exception:
-            self.get_session().close()
-            raise
 
         if result is None:
             return result
 
-        if not isinstance(result, (list, tuple)):
-            result = [result]
+        # we discard the first item of the result row,
+        # which was what the query was initialised with and not one of the requested projection (see self._build)
+        result = result[1:]
 
-        if not self._projection_index_to_field or len(result) != len(self._projection_index_to_field):
-            raise Exception('length of query result does not match the number of specified projections')
+        if len(result) != self._requested_projections:
+            raise Exception(
+                f'length of query result ({len(result)}) does not match '
+                f'the number of specified projections ({self._requested_projections})'
+            )
 
-        return [self.to_backend(rowitem) for rowitem in result]
+        return [self.to_backend(r) for r in result]
 
     def iterall(self, data: QueryDictType, batch_size: Optional[int]) -> Iterable[List[Any]]:
         """Return an iterator over all the results of a list of lists."""
-        query = self.update_query(data)
-        tag_to_index_dict = self._projection_index_to_field
-        try:
-            if not tag_to_index_dict:
-                raise Exception(f'Got an empty dictionary: {tag_to_index_dict}')
-
-            results = query.yield_per(batch_size)
+        with self.use_query(data) as query:
 
-            if len(tag_to_index_dict) == 1:
-                # Sqlalchemy, for some strange reason, does not return a list of lsits
-                # if you have provided an ormclass
-
-                if list(tag_to_index_dict.values()) == ['*']:
-                    for rowitem in results:
-                        yield [self.to_backend(rowitem)]
-                else:
-                    for rowitem, in results:
-                        yield [self.to_backend(rowitem)]
-            elif len(tag_to_index_dict) > 1:
-                for resultrow in results:
-                    yield [self.to_backend(rowitem) for rowitem in resultrow]
-            else:
-                raise ValueError('Got an empty dictionary')
-        except Exception:
-            self.get_session().close()
-            raise
+            for resultrow in query.yield_per(batch_size):  # type: ignore[arg-type] # pylint: disable=not-an-iterable
+                # we discard the first item of the result row,
+                # which is what the query was initialised with
+                # and not one of the requested projection (see self._build)
+                resultrow = resultrow[1:]
```

Comment:
This logic occurs in multiple places. Can we extract this somehow, maybe in the `yield_per` method? I must say I don't understand why this is necessary and it looks really weird.

Reply:
as in discarding the first item? The only other place it occurs is just above, in `def first`.
The reason is what I explain in the initial comment of the PR (and the start of `def _build`); you have to initialise a query with "something" (an entity or column) which then gets returned, but this is not a projection that was actually requested from the QueryBuilder, so we discard it

Reply:
I understand why it has to be removed because it was added artificially to initialize the query. What I don't understand is why we need to initialize it with an entity that is not actually part of the query. I guess the misunderstanding is more with the `_build` part and not so much the result fetching part. If you call `QueryBuilder().append(Node)` why is the `Node` not the only entity added in the query? Why does it need to be initialized with `Node` before `Node` being added again?

Reply:
Say you did this:

```python
QuerBuilder().append(Node).append(Group, project="label").all()
```

What you ideally want (simplifying without join tables) is:

```sql
SELECT Group.label
FROM Node
INNER JOIN Group ON Node.id = Group.node_id;
```

But in sqlalchemy, with the current way we construct a query, we need to do something like this:

```python
# initialise
query = Query(Node.id)
# add joins
query = query.join(Group, Node.id == Group.node_id)
# add projections
query = query.add_columns(Group.label)
```

and you end up with

```sql
SELECT Node.id, Group.label
FROM Node
INNER JOIN Group ON Node.id = Group.node_id;
```

We don't need the `Node.id`, but we have to initialise the query with an entity to join to 🤷 

Now it may be possible to completely restructure how you build the query (particularly with the v2 `select` API), to somehow work out everything you need first, and end up with something like:

```python
select(Node.id, Group.label).join(Group, Node.id == Group.node_id)
```

But I think that is beyond the scope of this PR, and would be tricky for the more complex joins

Make sense?

Reply:
Note, if you do: `QueryBuilder().append(Node)`, then I believe the actual query resulting from

```python
query = Query(Node.id)
query = query.add_columns(Node)
```

will actually look the same, since sqlalchemy will simply de-duplicate the fact that you are essentially asking for the id twice:

```sql
SELECT Node.id, Node.uuid, ... From Node
```

then reform the result from the database as `(Node.id, Node)`, i.e. there is no loss in performance of querying the database for this simple query

Reply:
Thanks for the write-up. That makes sense.